### ISSUE TITLE
[MODEL] XGBoost con optimización Optuna

### ISSUE THREAD

**Issue by JonnyBP** [Bootcamp-IA-P6/Project_8_Equipo1_DataScientist#5]
Entrenar XGBClassifier. 

- Usar Optuna para optimizar: n_estimators (50-500), max_depth (3-10), learning_rate (0.01-0.3), scale_pos_weight (para desbalance). 
- Definir la función objetivo de Optuna usando AUC-ROC con StratifiedKFold.
- Ejecutar mínimo 50 trials. 
- Guardar el mejor modelo en models/xgb_best.joblib. 
- Crear configs/experiments/exp_003_xgboost.yaml con los mejores hiperparámetros encontrados.

**Comment by gemayc** [comment]
## 📋 Descripción

Implementación de modelos de Machine Learning para predicción de accidentes cerebrovasculares (stroke) usando **Logistic Regression** y **XGBoost** con tracking completo en **MLflow** y optimización de hiperparámetros mediante **Optuna**.

## 🎯 Cambios Principales

### 1. **Preparación de Datos**
- Limpieza basada en análisis exploratorio (EDA)
  - Menores de 16 años con `smoking_status='Unknown'` → `'not_applied'`
  - `work_type='children'` → `'not_applied'`
- Dos versiones del dataset:
  - **FULL**: 4,981 pacientes (incluye niños)
  - **ADULTS**: 4,158 pacientes (solo edad > 16)

### 2. **Pipeline Automático**
Implementación de dos funciones reutilizables:

**`train_logistic_with_pipeline()`**
- Pipeline: OneHotEncoder → StandardScaler → SMOTE → LogisticRegression
- Detecta overfitting (compara Train vs Test)
- Retorna: pipeline + 8 métricas (accuracy, precision, recall, f1, auc, diff_acc, diff_f1)

**`train_xgboost_with_pipeline()`**
- Pipeline: OneHotEncoder → SMOTE → XGBoost (sin scaling, no lo necesita)
- Mismas métricas que Logistic

### 3. **MLflow: Tracking Completo**
Registra para cada experimento:
- **Parámetros**: Configuración del modelo
- **Métricas**: accuracy, precision, recall, f1, auc, overfitting_diff_accuracy, overfitting_diff_f1
- **Artefactos**: Matrices de confusión (PNG)
- **Run Name**: Formato `{MODELO}_{DATASET}_{VERSION}`

**Total: 10 experimentos rastreados**
```
- 5 entrenamientos iniciales (sin optimizar)
- 5 entrenamientos optimizados (post-Optuna)
```

### 4. **Optuna: Optimización de Hiperparámetros**

**Logistic Regression**
```python
def objective_logistic_generic(trial, X_train, X_test, y_train, y_test, dataset_version):
    C = trial.suggest_float('C', 0.001, 10.0, log=True)
    penalty = trial.suggest_categorical('penalty', ['l2', 'l1'])
    # Entrena y devuelve F1-Score
```

**XGBoost**
```python
def objective_xgboost_generic(trial, X_train, X_test, y_train, y_test, dataset_version):
    n_estimators = trial.suggest_int('n_estimators', 50, 300)
    max_depth = trial.suggest_int('max_depth', 3, 10)
    learning_rate = trial.suggest_float('learning_rate', 0.01, 0.3)
    # Entrena y devuelve F1-Score
```

Ejecuta **20 intentos por versión** → encuentra mejores parámetros automáticamente.

### 5. **Detección de Overfitting**

Se calcula dentro de cada entrenamiento:
```python
diff_acc = abs(accuracy_train - accuracy_test)
diff_f1 = abs(f1_train - f1_test)

# Registra en MLflow
mlflow.log_metric("overfitting_diff_accuracy", diff_acc)
mlflow.log_metric("overfitting_diff_f1", diff_f1)
```

**Resultados observados:**
- Logistic FULL: ~5.6% (leve overfitting) ⚠️
- XGBoost FULL: ~0.5% (sin overfitting) ✅

### 6. **Matrices de Confusión**
Se genera una matriz para cada entrenamiento:
- Guardadas en carpeta `assets/`
- Registradas en MLflow como artefactos
- Permiten visualizar falsos positivos y negativos

## 📊 Resultados Clave

### Logistic Regression (FULL, optimizado)
```
Accuracy:  0.7482
Precision: 0.9423
Recall:    0.7482
F1:        0.8185
AUC:       0.8372
```

### XGBoost (FULL, sin optimizar)
```
Accuracy:  0.9278
Precision: 0.9188
Recall:    0.9278
F1:        0.9231
AUC:       0.7885
```

**Observación**: XGBoost tiene mejor accuracy pero Logistic tiene mejor balance (F1).

## 🗂️ Estructura del Proyecto

```
proyecto/
├── data/
│   └── raw/
│       └── stroke_dataset.csv
├── notebooks/
│   ├── 01_EDA.ipynb                    (análisis exploratorio)
│   └── 02_Modeling_v2.ipynb            (entrenamientos + Optuna + MLflow)
├── assets/
│   ├── confusion_matrix_lr_full.png
│   ├── confusion_matrix_lr_adults.png
│   ├── confusion_matrix_xgb_full.png
│   ├── confusion_matrix_xgb_adults.png
│   ├── confusion_matrix_lr_full_opt.png
│   ├── confusion_matrix_lr_adults_opt.png
│   ├── confusion_matrix_xgb_full_opt.png
│   └── confusion_matrix_xgb_adults_opt.png
├── mlruns/                             (experimentos MLflow - NO SUBIR)
├── .gitignore
└── README.md
```

## 🚀 Cómo Usar

### Ejecutar el notebook completo
```bash
jupyter notebook notebooks/02_Modeling_v2.ipynb
```

### Ver experimentos en MLflow
```bash
mlflow ui
# Abre http://localhost:5000
```

### Ver parámetros optimizados
En el notebook, busca:
```python
print(best_params_logistic_full)
print(best_params_logistic_adults)
print(best_params_xgb_full)
print(best_params_xgb_adults)
```

## 📝 Archivos Modificados

- ✅ `notebooks/02_Modeling_v2.ipynb` - Nuevo notebook con ML completo
- ✅ `assets/` - Matrices de confusión
- ✅ `.gitignore` - Excluye mlruns/ y archivos temporales


## 🔧 Librerías Utilizadas

```
pandas, numpy, scikit-learn, xgboost, imbalanced-learn, optuna, mlflow, matplotlib, seaborn
```

## ✅ Testing / Validación

- ✓ Ambos modelos entrenan sin errores
- ✓ MLflow registra 10 experimentos correctamente
- ✓ Optuna encuentra parámetros válidos (20 intentos c/u)
- ✓ Matrices de confusión generadas correctamente
- ✓ Métricas calculadas (accuracy, precision, recall, f1, auc)
- ✓ Overfitting detectado y registrado

## 📚 Lecciones Aprendidas

1. **SMOTE dentro del Pipeline** evita data leakage automáticamente
2. **Scaling es necesario para Logistic**, no para XGBoost (tree-based)
3. **F1-Score es mejor que Accuracy** para datos desbalanceados (95% vs 5%)
4. **MLflow permite reproducibilidad** y comparación fácil entre modelos
5. **Optuna automatiza búsqueda** de mejores parámetros
